ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -483,22 +487,74 @@ public void hardDeleteTest() throws InterruptedException, IOException, StoreExce
-      LogSegment segment = log.getSegment(offset.getName());
-      long size = putValue.getSize() - HARD_DELETE_START_OFFSET - HARD_DELETE_LAST_PART_SIZE;
-      ByteBuffer readBuf = ByteBuffer.allocate((int) size);
-      segment.readInto(readBuf, offset.getOffset() + HARD_DELETE_START_OFFSET);
-      readBuf.flip();
-      while (readBuf.hasRemaining()) {
-        assertEquals("Hard delete has not zeroed out the data", (byte) 0, readBuf.get());
-      }
+
+    long expectedProgress = log.getDifference(logOrder.lastKey(), new Offset(log.getFirstSegment().getName(), 0));
+    assertEquals("Hard delete did not make expected progress", expectedProgress, index.hardDeleter.getProgress());
+    verifyEntriesForHardDeletes(deletedKeys, true);
+  }
+
+  /**
+   * Tests that hard delete pause and resume where in hard deletes is done by the daemon thread and not explicitly
+   * invoked by the tests
+   * @throws InterruptedException
+   * @throws IOException
+   * @throws StoreException
+   */
+  @Test
+  public void hardDeletePauseResumeTest() throws InterruptedException, IOException, StoreException {
+    properties.put("store.deleted.message.retention.days", Integer.toString(1));
+    properties.put("store.hard.delete.bytes.per.sec", Integer.toString(Integer.MAX_VALUE / 10));
+    properties.setProperty("store.data.flush.interval.seconds", "3600");
+    // enable daemon thread to run hard deletes
+    properties.put("store.enable.hard.delete", "true");
+    reloadIndex(false);
+    assertTrue("Hard delete is not running", index.hardDeleter.isRunning());
+    // IndexSegment still uses real time so advance time so that it goes 2 days past the real time.
+    advanceTime(SystemTime.getInstance().milliseconds() + 2 * Time.MsPerSec * Time.SecsPerDay);
+    long expectedProgress = log.getDifference(logOrder.lastKey(), new Offset(log.getFirstSegment().getName(), 0));
+    // give it some time so that hard delete completes one cycle
+    waitUntilExpectedProgress(expectedProgress, 1000);
+    verifyEntriesForHardDeletes(deletedKeys, true);
+
+    index.hardDeleter.pause();
+    assertTrue("Hard deletes should have been paused ", index.hardDeleter.isPaused());
+    advanceTime(HARD_DELETE_SLEEP_TIME_WHEN_CAUGHT_UP_MS);
+    waitUntilExpectedState(Thread.State.WAITING, HARD_DELETE_SLEEP_TIME_WHEN_CAUGHT_UP_MS + 1, 10);
+
+    // delete two entries
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    Set<MockId> idsToDelete = new HashSet<>();
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    for (MockId id : idsToDelete) {
+      addDeleteEntry(id);
     }
+    advanceTime(SystemTime.getInstance().milliseconds() + 2 * Time.MsPerSec * Time.SecsPerDay);
+    waitUntilExpectedState(Thread.State.WAITING, HARD_DELETE_SLEEP_TIME_WHEN_CAUGHT_UP_MS + 1, 10);
```

Comment:
just to ensure that if there are new deleted entries that are eligible to be hard deleted, the hard delete doesn't wake up to go about hard deleting it.